logging/google/cloud/logging/handlers/transports/background_thread.py

```diff
@@ -19,137 +19,224 @@
+                    # back before quitting.
+                    continue
+
+                batch.log_struct(**item)
+
+            total_logs = len(batch.entries)
+
+            try:
+                if total_logs:
+                    batch.commit()
+                    _LOGGER.debug('Submitted %d logs', total_logs)
+            except Exception:
+                _LOGGER.error(
+                    'Failed to submit %d logs.', total_logs, exc_info=True)
+            finally:
+                # Mark all collected tasks as done.
+                for n in range(len(items)):
+                    self._queue.task_done()
+
+            if quit:
+                break
+
+        _LOGGER.debug('Background thread exited gracefully.')
+
+    def start(self):
+        """Starts the background thread.
+
+        Additionally, this registers a handler for process exit to attempt
+        to send any pending log entries before shutdown.
         """
-        try:
-            self._entries_condition.acquire()
+        with self._operational_lock:
+            if self.is_alive:
+                return
+
             self._thread = threading.Thread(
-                target=self._run, name=_WORKER_THREAD_NAME)
+                target=self._thread_main,
+                name=_WORKER_THREAD_NAME)
             self._thread.setDaemon(True)
             self._thread.start()
-        finally:
-            self._entries_condition.release()
-            atexit.register(self._stop)
+            atexit.register(self._main_thread_terminated)
+
+    def stop(self, grace_period=None):
+        """Signals the background thread to stop.
 
-    def _stop(self):
-        """Signals the worker thread to shut down
+        This does not terminate the background thread. It simply queues the
+        stop signal. If the main process exits before the background thread
+        processes the stop signal, it will be terminated without finishing
+        work. The :param:`grace_period` parameter will give the background
+        thread some time to finish processing before this function returns.
 
-        Also waits for ``stop_timeout`` seconds for the worker to finish.
+        :type grace_period: float
+        :param grace_period: If specified, this method will block up to this
+            many seconds to allow the background thread to finish work before
+            returning.
 
-        This method is called by the ``atexit`` handler registered by
-         :meth:`start`.
+        :rtype: bool
+        :returns: True if the thread terminated. False if the thread is still
+            running.
         """
-        if not self.started or self.stopping:
-            return
+        if not self.is_alive:
+            return True
+
+        with self._operational_lock:
+            self._queue.put_nowait(_WORKER_TERMINATOR)
+
+            if grace_period is not None:
+                print('Waiting up to %d seconds.' % grace_period)
```

Comment:
I meant to print.

I'd love to use format, but I'm using `%` to be consistent with the rest of the code.